File: android/versioned-abis/expoview-abi29_0_0/src/main/java/abi29_0_0/host/exp/exponent/modules/api/components/barcodescanner/BarCodeScannerViewManager.java
package abi29_0_0.host.exp.exponent.modules.api.components.barcodescanner;

import android.support.annotation.Nullable;

import abi29_0_0.com.facebook.react.bridge.ReadableArray;
import abi29_0_0.com.facebook.react.common.MapBuilder;
import abi29_0_0.com.facebook.react.uimanager.ThemedReactContext;
import abi29_0_0.com.facebook.react.uimanager.ViewGroupManager;
import abi29_0_0.com.facebook.react.uimanager.annotations.ReactProp;
import abi29_0_0.host.exp.exponent.modules.api.components.barcodescanner.BarCodeScannerView.Events;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;

public class BarCodeScannerViewManager extends ViewGroupManager<BarCodeScannerView> {
    private static final String REACT_CLASS = "ExponentBarCodeScanner";

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    public BarCodeScannerView createViewInstance(ThemedReactContext themedReactContext) {
        return new BarCodeScannerView(themedReactContext);
    }

    @Override
    @Nullable
    public Map getExportedCustomDirectEventTypeConstants() {
        MapBuilder.Builder builder = MapBuilder.builder();
        for (Events event : Events.values()) {
            builder.put(event.toString(), MapBuilder.of("registrationName", event.toString()));
        }
        return builder.build();
    }


    @ReactProp(name = "type")
    public void setType(BarCodeScannerView view, int type) {
        view.setCameraType(type);
    }

    @ReactProp(name = "torchMode")
    public void setTorchMode(BarCodeScannerView view, int torchMode) {
        view.setTorchMode(torchMode);
    }

    @ReactProp(name = "barCodeTypes")
    public void setBarCodeTypes(BarCodeScannerView view, ReadableArray barCodeTypes) {
        if (barCodeTypes == null) {
            return;
        }
        List<Integer> result = new ArrayList<>(barCodeTypes.size());
        for (int i = 0; i < barCodeTypes.size(); i++) {
            result.add(barCodeTypes.getInt(i));
        }
        view.setBarCodeTypes(result);
    }
}

File: android/versioned-abis/expoview-abi47_0_0/src/main/java/abi47_0_0/expo/modules/barcodescanner/BarCodeScannerViewManager.kt
package abi47_0_0.expo.modules.barcodescanner

import android.content.Context
import abi47_0_0.expo.modules.core.ModuleRegistry
import abi47_0_0.expo.modules.core.ModuleRegistryDelegate
import abi47_0_0.expo.modules.core.ViewManager
import abi47_0_0.expo.modules.core.interfaces.ExpoProp
import abi47_0_0.expo.modules.interfaces.barcodescanner.BarCodeScannerSettings
import java.util.*

class BarCodeScannerViewManager(
  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
) : ViewManager<BarCodeScannerView>() {

  enum class Events(private val mName: String) {
    EVENT_ON_BAR_CODE_SCANNED("onBarCodeScanned");

    override fun toString() = mName
  }

  override fun getName() = TAG

  override fun onCreate(moduleRegistry: ModuleRegistry) {
    moduleRegistryDelegate.onCreate(moduleRegistry)
  }

  override fun createViewInstance(context: Context) =
    BarCodeScannerView(context, moduleRegistryDelegate)

  override fun getViewManagerType() = ViewManagerType.GROUP

  override fun getExportedEventNames() =
    Events.values().map { it.toString() }

  @ExpoProp(name = "type")
  fun setType(view: BarCodeScannerView, type: Int) {
    view.setCameraType(type)
  }

  @ExpoProp(name = "barCodeTypes")
  fun setBarCodeTypes(view: BarCodeScannerView, barCodeTypes: ArrayList<Double?>?) {
    if (barCodeTypes != null) {
      val settings = BarCodeScannerSettings().apply {
        putTypes(barCodeTypes)
      }
      view.setBarCodeScannerSettings(settings)
    }
  }

  companion object {
    private const val TAG = "ExpoBarCodeScannerView"
  }
}
